### PR TITLE
Feature 65611

### PR DESCRIPTION
**Pull request by cse-peter-account**
This pull request implements Feature #65611 (https://bz.apache.org/bugzilla/show_bug.cgi?id=65611).

According to step 2 of the following URL, JMeter only accepts IPv4 addresses when specifying a remote worker node.

https://jmeter.apache.org/usermanual/remote-test.html

In this scenario, entering an IPv6 address results in a NumberFormatException.  The purpose of this pull request is to eliminate this NumberFormatException.

### CODE REVIEW

File: src/core/src/main/java/org/apache/jmeter/engine/ClientJMeterEngine.java [hunk 1]
```diff
@@ -58,12 +59,18 @@ private static RemoteJMeterEngine getEngine(String hostAndPort)
...
+
+        if (!isIPv6Address(host)){
+            int indexOfSeparator = hostAndPort.indexOf(':');
+            if (indexOfSeparator >= 0) {
+                host = hostAndPort.substring(0, indexOfSeparator);
+                String portAsString = hostAndPort.substring(indexOfSeparator+1);
+                port = Integer.parseInt(portAsString);
+            }
+        } else {
+            System.out.println("IPv6 address detected.  No parsing is required.");
```
Comment (ham1) [inline src/core/src/main/java/org/apache/jmeter/engine/ClientJMeterEngine.java:71]: Should this else block and sout be removed?

Comment (cse-peter-account) [reply src/core/src/main/java/org/apache/jmeter/engine/ClientJMeterEngine.java]: Done:

commit 2fd1f1eb9057ba24ace59fc3b0b5452001aa1a40

Please kindly review.  Thanks!

**Comment by FSchumacher** [comment]
What happens, when a user wants to use IPv6 and a port?
Is it worth adding a complete library, to check for a IPv6 address, only?

**Comment by cse-peter-account** [comment]
> What happens, when a user wants to use IPv6 and a port? Is it worth adding a complete library, to check for a IPv6 address, only?

************************************************************************************
If a user wants to use an IPv6 port, it shouldn't be necessary to include a port, according to the following URL:

https://docs.oracle.com/javase/8/docs/technotes/guides/net/ipv6_guide/#:~:text=IPv6%20in%20Java%20is%20transparent%20and%20automatic.%20Porting%20is%20not%20necessary

"IPv6 in Java is transparent and automatic. Porting is not necessary;"
************************************************************************************
I added a small, specialized library to check for IPv6, since that is what the library is specialized to do.  If we add manual code to check for IPv6 format, human errors could be introduced either now or later.

Is it mandatory to remove the library if we want to push this feature to apache:master?
************************************************************************************

Thanks!

**Comment by cse-peter-account** [comment]
> > What happens, when a user wants to use IPv6 and a port? Is it worth adding a complete library, to check for a IPv6 address, only?
> 
> If a user wants to use an IPv6 port, it shouldn't be necessary to include a port, according to the following URL:
> 
> https://docs.oracle.com/javase/8/docs/technotes/guides/net/ipv6_guide/#:~:text=IPv6%20in%20Java%20is%20transparent%20and%20automatic.%20Porting%20is%20not%20necessary
> 
> "IPv6 in Java is transparent and automatic. Porting is not necessary;"
> 
> I added a small, specialized library to check for IPv6, since that is what the library is specialized to do. If we add manual code to check for IPv6 format, human errors could be introduced either now or later.
> 
> Is it mandatory to remove the library if we want to push this feature to apache:master?
> 
> Thanks!

**Comment by FSchumacher** [comment]
I think, you misread the docs. The *porting* does not mean the *port* part, but the recompilation/adaption of the java source, when using the Java API with IPv6.
What I meant is, that currently we can give JMeter an IP address plus a port in one string ("1.2.3.4:80" for IPv4 1.2.3.4 with port 80), which as you found out is problematic, when you specify an IPv6 address (as it contains colons, which are used currently to separate ip and port). Your solution drops the port part completely, which is bad (in my opinion) and probably unneeded.
What I have seen is, that if you want to specify an IPv6 plus port combination, you put the IPv6 address in square brackets like this: "[0::1]:80". If we use this standard notation, the parsing will be simplified to looking at the first non space char. If it is a opening square bracket ('['), we know, that the user wants to specify an IPv6 address. We can look for the closing bracket extract the IPv6 and are happy. (As a side effect, we can drop adding the new library and ask the normal Java API to confirm the validity of the IPv6 address)

**Comment by FSchumacher** [comment]
Thanks for the feature request and the initial implementation. It would be nice, if you could check next nightly and report back (probably better on bugzilla), if it works for you.

**Pull request closed without merging**